src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs

```diff
@@ -0,0 +1,358 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.InteropServices;
+
+internal partial class Interop
+{
+    internal partial class mincore
+    {
+        internal const int COR_E_PLATFORMNOTSUPPORTED = unchecked((int)0x80131539);
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/bb762188.aspx
+        [DllImport(Libraries.ShellFolders, CharSet = CharSet.Unicode, SetLastError = false, BestFitMapping = false, ExactSpelling = true, EntryPoint = "SHGetKnownFolderPath")]
+        private static extern int SHGetKnownFolderPath_ShellFolders(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            uint dwFlags,
+            IntPtr hToken,
+            out string ppszPath);
+
+        private static bool s_skipShellFolders;
+
+// Disabling the warning about availability.
+#pragma warning disable BCL0015
+        [DllImport(Libraries.Shell, CharSet = CharSet.Unicode, SetLastError = false, BestFitMapping = false, ExactSpelling = true, EntryPoint = "SHGetKnownFolderPath")]
+        private static extern int SHGetKnownFolderPath_Shell32(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            uint dwFlags,
+            IntPtr hToken,
+            out string ppszPath);
+#pragma warning restore BCL0015
+
+        internal static int SHGetKnownFolderPath(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            uint dwFlags,
+            IntPtr hToken,
+            out string ppszPath)
+        {
+            // The ShellFolders API set isn't available on Win7, while the direct Shell32 isn't available
+            // on some mincore versions of Windows. We'll attempt the newer API set version and fall back
+            // to the older "Shell32" if needed.
+            if (!s_skipShellFolders)
+            {
+                try
+                {
+                    return SHGetKnownFolderPath_ShellFolders(rfid, dwFlags, hToken, out ppszPath);
+                }
+                catch (TypeLoadException)
```

Comment:
I see ... I have not realized that these exceptions inherit from TypeLoadException.

> Why is it questionable? 

- There is annoying exception thrown that shows up in the debugger. I have seen like dozen cases over the years in the full framework where this exact pattern was used, only to be replaced with LoadLibrary/GetProcAddress later.
- It is not compatible with AOT compilation.

> we could use LoadLibrary/GetProcAddress in managed for all of our Windows distributions

The only place where LoadLibrary/GetProcAddress cannot be used today is UWP. This implementation cannot be used for it anyway because it is not compatible with UWP AOT toolchain, and the PInvoke to shell32 is not allowed there either (it does not matter that it should not be ever executed).